lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -1469,29 +1511,119 @@ void LowerTypesPass::runOnOperation() {
   // Lower each module and return a list of Nlas which need to be updated with
   // the new symbol names.
   SmallVector<NlaNameNewSym> nlaToNewSymList;
-  SmallVector<InnerRefRecord> instanceSymNames;
+  SmallVector<InnerRefRecord> opSymNames;
   std::mutex nlaAppendLock;
-  auto lowerModules = [&](auto op) {
+  // This lambda, executes in parallel for each Op within the circt.
+  auto lowerModules = [&](Operation *op) -> void {
     auto tl = TypeLoweringVisitor(&getContext(), flattenAggregateMemData,
                                   preserveAggregate, preservePublicTypes,
                                   symTbl, cache);
     tl.lowerModule(op);
     std::lock_guard<std::mutex> lg(nlaAppendLock);
+    // This section updates shared data structures using a lock.
     nlaToNewSymList.append(tl.getNLAs());
-    // Record all the instances in the module that have a symbol.
-    if (auto mod = dyn_cast<FModuleOp>(op))
-      for (auto instName : tl.getInstanceSymNames()) {
-        instanceSymNames.emplace_back(mod.getNameAttr(), instName.getFirst(),
-                                      instName.getSecond());
-      }
+    StringAttr modName;
+    ArrayAttr portSyms;
+    // Create an inner_sym to op record, which will be used to lookup ops given
+    // the InnerRefAttr, during NLA fixup.
+    if (FModuleOp mod = dyn_cast<FModuleOp>(op)) {
+      modName = mod.getNameAttr();
+      portSyms = mod.getPortSymbolsAttr();
+    } else if (auto mod = dyn_cast<FExtModuleOp>(op)) {
+      modName = mod.getNameAttr();
+      portSyms = mod.getPortSymbolsAttr();
+    } else
```

Comment:
I believe you can combine this by casting to an `FModuleLikeOp` instead of one or the other.

Reply:
I think avoiding `FModuleLike` might be a bit more efficient.
Andrew Young had mentioned that, `auto mod = dyn_cast<FModuleLike>(op)` ends up being a hashtable lookup, and might not be worth saving few extra LOC.